include/swift/Driver/ExperimentalDependencyDriverGraph.h

```diff
@@ -0,0 +1,338 @@
+/// without changing as much code.
+class DependencyGraphImpl {
+public:
+  /// Use the status quo LoadResult for now.
+  using LoadResult = typename swift::DependencyGraphImpl::LoadResult;
+};
+
+//==============================================================================
+// MARK: DriverGraph
+//==============================================================================
+
+/// See \ref Node in ExperimentalDependencies.h
+class DriverGraph {
+
+  /// Find nodes, first by the swiftDeps file, then by key.
+  /// Supports searching specific files for a node matching a key.
+  /// Such a search is useful when integrating nodes from a given source file to
+  /// see which nodes were there before integration and so might have
+  /// disappeared.
+  ///
+  /// Some nodes are in no file, for instance a dependency on a Decl in a source
+  /// file whose swiftdeps has not been read yet. For these, the filename is the
+  /// empty string.
+  ///
+  /// Don't add to this collection directly; use \ref addToMap
+  /// instead because it enforces the correspondence with the swiftFileDeps
+  /// field of the node.
+  /// TODO: Fix above comment
+  ///
+  /// Sadly, cannot use an optional string for a key.
+  using NodeMap =
+      BiIndexedTwoStageMap<std::string, DependencyKey, DriverNode *>;
+  NodeMap nodeMap;
+
+  /// Since dependency keys use baseNames, they are coarser than individual
+  /// decls. So two decls might map to the same key. Index the second level by
+  /// file. Very useful to find a Node given the key, for instance when finding
+  /// all the nodes affected by a change in one.
+
+  /// Tracks def-use relationships by DependencyKey.
+  std::unordered_map<DependencyKey, std::unordered_set<DependencyKey>>
+      usesByDef;
+
+  // Supports requests from the driver to getExternalDependencies.
+  std::unordered_set<std::string> externalDependencies;
+
+  /// The new version of "Marked."
+  /// Record cascading jobs by swiftDepsFilename because that's what
+  /// nodes store directly.
+  ///
+  /// The status quo system uses "cascade" for the following:
+  /// Def1 -> def2 -> def3, where arrows are uses, so 3 depends on 2 which
+  /// depends on 1. The first use is said to "cascade" if when def1 changes,
+  /// def3 is dirtied.
+  /// TODO: Move cascadingJobs out of the graph, ultimately.
+  /// If marked, any Job that depends on me must be rebuilt after compiling me
+  /// if I have changed.
+
+  std::unordered_set<std::string> cascadingJobs;
+
+  /// Keyed by swiftdeps filename, so we can get back to Jobs.
+  std::unordered_map<std::string, const Job *> jobsBySwiftDeps;
+
+  /// For debugging, the driver writes out a dot file of the graph every time a
+  /// Frontend swiftdeps is read and integrated. In order to keep subsequent
+  /// files for the same job distinct, keep a sequence number for each job.
+  std::unordered_map<const Job *, uint> dotFileSequenceNumberByJob;
```

Comment:
Nitpick: I don't know where the `uint` comes from but it's not standard C++ or LLVM. (We generally use `unsigned` for that particular C type.)

Reply:
Thanks. Done.